### PR TITLE
Fix #10217: Avoid exponential behavior in derivesFrom

### LINKED ISSUE #10217: typing union in type parameter argument grows exponentially
## Minimized code

```Scala
trait A
trait B
trait C
trait D
trait E
trait F
trait G
trait H
trait I
trait J
trait K
trait L
trait M
trait N
trait O
trait P
trait Q
trait R
trait S
trait T
trait U
trait V
trait W

class Foo[T]

val f = Foo[A | B | C | D | E | F | G | H | I | J | K | L | M | N | O | P | Q | R | S | T | U]
```

## Output

With Dotty `0.27.0-RC1` through `0.28.0-bin-20201014-ca67e4d-NIGHTLY`, a file containing just the above code compiles in less than a second on my laptop. **However, using `3.0.0-M1` and later (including the latest nightly build, `3.0.0-M2-bin-20201105-c20be38-NIGHTLY`), the above code takes 287 seconds to compile:**

```scala
sbt:dotty-simple> compile
[info] Compiling 1 Scala source to C:\Git\Overload\target\scala-3.0.0-M1\classes ...
[success] Total time: 287 s (04:47), completed 6 Nov 2020, 21:37:04
```

Instantiating `Foo` with a union of only types `A` through `T` (i.e. omitting `U`) reduces the compile time to 103 seconds:

```scala
sbt:dotty-simple> compile
[info] Compiling 1 Scala source to C:\Git\Overload\target\scala-3.0.0-M1\classes ...
[success] Total time: 103 s (01:43), completed 6 Nov 2020, 21:40:16
```

Instantiating `Foo` with a union of only types `A` through `S` reduces the compile time to 35 seconds.

Instantiating `Foo` with a union of only types `A` through `R` reduces the compile time to 11 seconds.

Instantiating `Foo` with a union of only types `A` through `Q` reduces the compile time to 4 seconds.

Instantiating `Foo` with a union of only types `A` through `P` reduces the compile time to 2 seconds.

## Expectation

The time required to compile code like this should not explode as terms are added to the union of types.

### PR DESCRIPTION
**Pull request by odersky**

### CODE REVIEW

**Comment by odersky** [comment]
`isBottomType` calls `derivesFom` again, which caused exponential behavior.

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 1]
```diff
@@ -242,8 +242,9 @@ object Types {
           // If the type is `T | Null` or `T | Nothing`, and `T` derivesFrom the class,
           // then the OrType derivesFrom the class. Otherwise, we need to check both sides
           // derivesFrom the class.
-          if defn.isBottomType(tp.tp1) then loop(tp.tp2)
-          else loop(tp.tp1) && (defn.isBottomType(tp.tp2) || loop(tp.tp2))
+          loop(tp.tp1) && loop(tp.tp2)
+          || tp.tp1.isNullType && loop(tp.tp2)
+          || tp.tp2.isNullType && loop(tp.tp2)
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/core/Types.scala:247]: Ther's a typo in here (`loop(tp.tp2)` should be `loop(tp.tp1)`), and this isn't quite the same thing as before; it doesn't handle `T | Nothing`, just `T | Null`. Instead of changing this I suggest changing isBottomType / isBottomTypeAfterErasure to use `tp.isRef` instead of `tp.derivesFrom` (you can't have a subclass of Null or Nothing anyway).

**Review by smarter — CHANGES_REQUESTED** [review]

File: compiler/src/dotty/tools/dotc/core/Types.scala (hunk 1 above)
Comment (odersky) [reply compiler/src/dotty/tools/dotc/core/Types.scala:247]: Changing `derivesFrom` to `isRef` does not work since for unions and intersections though.

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 2]
```diff
@@ -242,8 +242,9 @@ object Types {
           // If the type is `T | Null` or `T | Nothing`, and `T` derivesFrom the class,
           // then the OrType derivesFrom the class. Otherwise, we need to check both sides
           // derivesFrom the class.
-          if defn.isBottomType(tp.tp1) then loop(tp.tp2)
-          else loop(tp.tp1) && (defn.isBottomType(tp.tp2) || loop(tp.tp2))
+          loop(tp.tp1) && loop(tp.tp2)
+          || tp.tp1.isNullOrNothingType && loop(tp.tp2)
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/core/Types.scala:246]: `isBottomType` doesn't return true for `null` when `ctx.explicitNulls && !ctx.phase.erasedTypes` is true but this will, breaking null-safety.

**Review by smarter — CHANGES_REQUESTED** [review]

File: compiler/src/dotty/tools/dotc/core/Types.scala (hunk 2 above)
Comment (odersky) [reply compiler/src/dotty/tools/dotc/core/Types.scala]: Right. I changed it so that `isBottomType` no longer calls `derivesFrom`.

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 3]
```diff
@@ -242,8 +248,8 @@ object Types {
           // If the type is `T | Null` or `T | Nothing`, and `T` derivesFrom the class,
           // then the OrType derivesFrom the class. Otherwise, we need to check both sides
           // derivesFrom the class.
-          if defn.isBottomType(tp.tp1) then loop(tp.tp2)
-          else loop(tp.tp1) && (defn.isBottomType(tp.tp2) || loop(tp.tp2))
+          loop(tp.tp1) && (loop(tp.tp2) || defn.isBottomType(tp.tp2))
+          || defn.isBottomType(tp.tp1) && loop(tp.tp2)
```
Comment (noti0na1) [inline compiler/src/dotty/tools/dotc/core/Types.scala:252]: I think the behaviour of this is same as before, but if `tp = Null | (Null | (Null | ( ... | A)`, wouldn't `tp.derivesFrom(NullClass)` cause the expensive `loop(tp.tp2)` to run two times? So I think use `if` is better here.

File: tests/pos/i10217.scala [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+trait R
+trait S
+trait T
+trait U
+trait V
+trait W
+
+class Foo[T]
+
+val f = Foo[A | B | C | D | E | F | G | H | I | J | K | L | M | N | O | P | Q | R | S | T | U]
```
Comment (noti0na1) [inline tests/pos/i10217.scala:27]: Can we add another test `Null | (Null | (Null | ( ... | A)` here?

**Review by noti0na1 — CHANGES_REQUESTED** [review]

File: compiler/src/dotty/tools/dotc/core/Types.scala (hunk 3 above)
Comment (noti0na1) [reply compiler/src/dotty/tools/dotc/core/Types.scala:252]: Also, do we need to make sure `OrType(NullType, NothingType).derivesFrom(NothingClass) == false`?

**Comment by odersky** [comment]
I changed isBottomType to use a new method, which is linked with the `classSymbols` abstraction. I think that should be the right way to characterize it.

I also did some moving around and renaming to standardize the way we test for top and bottom types.

**Review by noti0na1 — APPROVED** [review]
